reps/2023-07-20-torch-trainer-apis.md

```diff
@@ -0,0 +1,371 @@
+
+With this, the user can directly interact with the Lightning interface, and is able define their training logic to use additional functionality such as `lightning.Trainer.test()`.
+
+```python
+import lightning
+import ray.train.lightning
+from ray.train.torch import TorchTrainer
+
+def train_func(config):
+
+    ...
+
+    devices = ray.train.lightning.get_devices()
+    strategy = ray.train.lightning.RayDDPStrategy()
+    environment = ray.train.lightning.RayEnvironment()
+
+    lightning_trainer = lightning.Trainer(
+        devices=devices,
+        strategy=strategy,
+        plugins=[environment],
+        **trainer_kwargs
+    )
+    ray.train.lightning.prepare_trainer(lightning_trainer)
+    lightning_trainer.fit(MyLightningModule(**module_kwargs), **fit_kwargs)
+
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_func,
+    ...
+)
+trainer.fit()
+```
+
+**Reporting & Checkpointing:**
+
+For Lightning, we introduce a `RayTrainReportCallback`(`Callback`) that will report metrics and checkpoints to Ray Train.
+
+```python
+from lightning.callbacks.pytorch import Checkpoint
+
+CHECKPOINT_FILE_NAME = "checkpoint.ckpt"
+
+class RayTrainReportCallback(Checkpoint):
+    # Define logic for calling `ray.train.report` as a Callback.
+    ...
+
+def train_func():
+    # Create checkpoint.
+    report_checkpoint_callback = RayTrainReportCallback()
+    trainer = Trainer(..., callbacks=[report_checkpoint_callback])
+    # Restore checkpoint.
+    checkpoint_path = None
+    checkpoint = ray.train.get_context().get_checkpoint()
+    if checkpoint:
+        checkpoint_dir = checkpoint.to_directory()
+        checkpoint_path = Path(checkpoint_dir) / CHECKPOINT_FILE_NAME
+    trainer.fit(model, ckpt_path=checkpoint_path, ...)
+```
+
+### HuggingFace Transformers
+
+#### Transformers
+```python 
+import transformers
+
+trainer = transformers.Trainer(**trainer_kwargs)
+trainer.train()
+```
+
+**`TransformersTrainer` (Current):**
+
+In the existing `TransformersTrainer`, we expose a thin `trainer_init_per_worker` function that allows the user to configure the `transformers.Trainer`.
+
+When `TransformersTrainer.fit` is called, it will run an opinionated training loop that constructs the `transformers.Trainer` and calls `transformers.Trainer.train()`.
+
+```python
+import transformers
+from ray.train.huggingface import TransformersTrainer
+
+def trainer_init_per_worker(train_dataset, eval_dataset, **config):
+    ...
+    return transformers.Trainer(**trainer_kwargs)
+
+trainer = TransformersTrainer(
+    trainer_init_per_worker=trainer_init_per_worker,
+    trainer_init_config=trainer_init_config,
+    ...
+)
+trainer.fit()
+```
+
+#### `TorchTrainer` (Proposed)
+
+In this proposal, we provide a `prepare_trainer` utility that the user can use directly in their training loop code to validate that the `Trainer` object is configured correctly to be compatible with Ray Train.
```

Comment:
For clarification, does the `prepare_trainer` only validate configuration or does it also patch some functionality/setup the environment etc?

Reply:
For flexibility of the API this may perform some modifications - updated the comment/example to reflect this.